Core/clim-basic/windowing/ports.lisp

```diff
@@ -43,34 +43,34 @@
             (t :clx-ttf))
     :null))
 
-(defun find-default-server-path ()
-  (loop for port in *server-path-search-order*
-        if (get port :port-type)
-           do (return-from find-default-server-path (list port))
-        finally (error "No CLIM backends have been loaded!")))
+(defgeneric find-port-type (port)
+  (:method ((port symbol))
+    (values (get port :port-type)
+            (get port :server-path-parser))))
 
 (defvar *all-ports* nil)
 
 (defun find-port (&key (server-path *default-server-path*))
-  (if (null server-path)
-      (setq server-path (find-default-server-path)))
-  (if (atom server-path)
-      (setq server-path (list server-path)))
-  (setq server-path
-        (funcall (get (first server-path) :server-path-parser) server-path))
-  (loop for port in *all-ports*
-        if (equal server-path (port-server-path port))
-        do (return port)
-        finally (let ((port-type (get (first server-path) :port-type))
-                      port)
-                  (if (null port-type)
-                      (error "Don't know how to make a port of type ~S"
-                             server-path))
-                  (setq port
-                        (funcall 'make-instance port-type
-                                 :server-path server-path))
-                  (push port *all-ports*)
-                  (return port))))
+  (unless (listp server-path)
+    (setf server-path (list server-path)))
+  (multiple-value-bind (port-class server-path)
+      (if (null server-path)
+          (loop for port in *server-path-search-order*
+                do (multiple-value-bind (class parser)
+                       (find-port-type port)
+                     (when class
+                       (return (values class (funcall parser (list class))))))
+                finally
+                   (error "No CLIM backends have been loaded!"))
+          (multiple-value-bind (class parser)
+              (find-port-type (first server-path))
```

Comment:
I think the error reporting in case this port does not exist is now missing.

Reply:
fixed